File: java/src/main/java/com/sasteval/vuln/easy/XssStoredDirect.java
package com.sasteval.vuln.easy;

import com.sasteval.util.DbUtil;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.io.PrintWriter;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.Statement;

/**
 * CWE-79: Cross-Site Scripting (Stored, Direct)
 * VULNERABILITY: Comments are stored in DB and rendered without encoding.
 */
public class XssStoredDirect extends HttpServlet {

    @Override
    protected void doPost(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {

        String comment = request.getParameter("comment");

        try (Connection conn = DbUtil.getConnection();
             PreparedStatement ps = conn.prepareStatement("INSERT INTO comments (comment) VALUES (?)")) {

            ps.setString(1, comment);
            ps.executeUpdate();

            response.setContentType("application/json");
            response.getWriter().println("{\"status\": \"stored\"}");

        } catch (Exception e) {
            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
            response.getWriter().println("{\"error\": \"" + e.getMessage() + "\"}");
        }
    }

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {

        response.setContentType("text/html");
        PrintWriter out = response.getWriter();

        out.println("<html><body><h1>Comments</h1>");

        try (Connection conn = DbUtil.getConnection();
             Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery("SELECT comment FROM comments")) {

            while (rs.next()) {
                // VULN: Stored comment rendered without HTML encoding
                out.println("<div>" + rs.getString("comment") + "</div>");
            }

        } catch (Exception e) {
            out.println("<p>Error: " + e.getMessage() + "</p>");
        }

        out.println("</body></html>");
    }
}

File: dotnet/Vuln/Easy/XssStoredDirect.cs
using System.Data.SQLite;
using System.Text;
using SastEval.Util;

namespace SastEval.Vuln.Easy;

/// <summary>
/// CWE-79: Cross-Site Scripting (Stored) - User comment stored in DB and rendered as HTML without encoding.
/// </summary>
public static class XssStoredDirect
{
    public static IResult HandlePost(HttpContext context)
    {
        var comment = context.Request.Query["comment"].ToString();

        if (string.IsNullOrEmpty(comment))
        {
            return Results.BadRequest("Missing 'comment' query parameter");
        }

        var conn = DbUtil.GetConnection();

        // VULNERABLE: Storing raw user input without sanitization
        using var cmd = new SQLiteCommand("INSERT INTO comments (body) VALUES (@body)", conn);
        cmd.Parameters.AddWithValue("@body", comment);
        cmd.ExecuteNonQuery();

        return Results.Ok("Comment stored");
    }

    public static IResult HandleGet(HttpContext context)
    {
        var conn = DbUtil.GetConnection();

        using var cmd = new SQLiteCommand("SELECT body FROM comments", conn);
        using var reader = cmd.ExecuteReader();

        var sb = new StringBuilder();
        sb.Append("<html><body><h1>Comments</h1>");

        while (reader.Read())
        {
            var body = reader["body"].ToString();
            // VULNERABLE: Rendering stored user input as raw HTML without encoding
            sb.Append($"<div class='comment'>{body}</div>");
        }

        sb.Append("</body></html>");

        return Results.Content(sb.ToString(), "text/html");
    }
}
